Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php

```diff
@@ -312,6 +314,68 @@ public function migrateMetaDataToWorkspaceName(\Closure $outputFn)
         $outputFn(sprintf('Migration applied to %s events.', count($this->eventsModified)));
     }
 
+    /**
+     *  Adds the "workspaceName" to the data of all content stream related events
+     *
+     *  Needed for feature "Add workspaceName to relevant events": https://github.com/neos/neos-development-collection/issues/4996
+     *
+     *  Included in May 2024 - before final Neos 9.0 release
+     *
+     * @param \Closure $outputFn
+     * @return void
+     */
+    public function migratePayloadToWorkspaceName(\Closure $outputFn): void
+    {
+
+        $backupEventTableName = DoctrineEventStoreFactory::databaseTableName($this->contentRepositoryId) . '_bkp_' . date('Y_m_d_H_i_s');
+        $outputFn('Backup: copying events table to %s', [$backupEventTableName]);
+        $this->copyEventTable($backupEventTableName);
+
+        $numberOfMigratedEvents = 0;
+        $workspaceNamesByContentStreamId = [];
+
+        foreach ($this->eventStore->load(VirtualStreamName::all()) as $eventEnvelope) {
+            $eventType = $eventEnvelope->event->type->value;
+            if (in_array($eventType, ['RootWorkspaceWasCreated', 'WorkspaceBaseWorkspaceWasChanged', 'WorkspaceWasCreated', 'WorkspaceWasDiscarded', 'WorkspaceWasRebased'], true)) {
+                $eventData = self::decodeEventPayload($eventEnvelope);
+                $workspaceNamesByContentStreamId[$eventData['newContentStreamId']] = $eventData['workspaceName'];
+                continue;
+            }
+            if (in_array($eventType, ['WorkspaceWasPartiallyPublished', 'WorkspaceWasPartiallyPublished', 'WorkspaceWasPublished'], true)) {
+                $eventData = self::decodeEventPayload($eventEnvelope);
+                $workspaceNamesByContentStreamId[$eventData['newSourceContentStreamId']] = $eventData['sourceWorkspaceName'];
```

Comment:
Its `WorkspaceWasPartiallyDiscarded::$newContentStreamId` not `WorkspaceWasPartiallyDiscarded::$newSourceContentStreamId` unlike in `WorkspaceWasPartiallyPublished`.

Reply:
Good catch! I checked the events multiple times but still got it wrong apparently :-/

Reply:
will take care

Reply:
`WorkspaceWasPartiallyDiscarded` wasn't handled at all, but it is _now_

Reply:
perfecto ❤️ i just compared it against my pr: https://github.com/neos/neos-development-collection/pull/5041
 where i already solved the riddle and there e2e pass. So with this addition it must be 100% correct.